resources/assets/lib/beatmaps/search-filter.tsx

```diff
@@ -0,0 +1,113 @@
+import { FilterKey } from 'beatmapset-search-filters';
+import { action, computed, makeObservable } from 'mobx';
+import { observer } from 'mobx-react';
+import core from 'osu-core-singleton';
+import * as React from 'react';
+import { classWithModifiers } from 'utils/css';
+import { formatNumber } from 'utils/html';
+import { FilterOption } from './available-filters';
+
+interface Props {
+  multiselect: boolean;
+  name: FilterKey;
+  options: FilterOption[];
+  title?: string;
+}
+
+@observer
+export class SearchFilter extends React.PureComponent<Props> {
+  private get controller() {
+    return core.beatmapsetSearchController;
+  }
+
+  @computed
+  private get currentSelection() {
+    const optionKeys = this.options.map((option) => option.id);
+
+    const filters = this.controller.getFilters(this.props.name)?.filter((option) => optionKeys.includes(option));
+    return new Set<string | null>(filters);
+  }
+
+  @computed
+  private get options() {
+    // normalize option keys
+    return this.props.options.map((option) => ({
+      id: option.id == null ? null : String(option.id),
+      name: option.name,
+    }));
+  }
+
+  constructor(props: Props) {
+    super(props);
+    makeObservable(this);
+  }
+
+  render() {
+    return (
+      <div className='beatmapsets-search-filter'>
+        {this.props.title != null && <span className='beatmapsets-search-filter__header'>{this.props.title}</span>}
+        <div className='beatmapsets-search-filter__items'>
+          {this.options.map(this.renderOption)}
+        </div>
+      </div>
+    );
+  }
+
+  private href(id: string | null) {
+    return this.controller.filters.href(this.props.name, this.newSelection(id) );
+  }
+
+  private isSelected(key: string | null) {
+    if (key == null && this.currentSelection.size === 0) return true;
+
+    return this.currentSelection.has(key);
+  }
+
+  // TODO: rename
+  private newSelection(key: string | null) {
+    const newSet = this.props.multiselect ? new Set(this.currentSelection) : new Set<string | null>();
+    if (this.isSelected(key)) {
+      newSet.delete(key);
+    } else {
+      newSet.add(key);
+    }
+
+    const value = [...newSet.values()].filter((x) => x != null).join('.');
+    return value.length === 0 ? null : value;
+  }
+
+  private readonly renderOption = (option: { id: string | null; name: string }) => {
+    const cssClasses = classWithModifiers('beatmapsets-search-filter__item', {
+      active: this.isSelected(option.id),
+      'featured-artists': option.id === 'featured_artists',
+    });
+
+    let text = option.name;
+    if (this.props.name === 'general' && option.id === 'recommended' && this.controller.recommendedDifficulty != null) {
+      text += ` (${formatNumber(this.controller.recommendedDifficulty, 2)})`;
+    }
+    // FIXME: do an actual navigation
+    return (
+      <a
+        key={option.id}
+        className={cssClasses}
+        data-filter-value={option.id}
+        href={this.href(option.id)}
```

Comment:
it looks like `q` parameter sometimes not updated correctly? `q=...` is stuck with old value until the filter is selected

Reply:
This is an existing bug that won't be fixed here as mention in the PR description

Reply:
looks like generating href using actual filters' `queryParams` is enough to fix it? Depending on the document href doesn't seem quite right anyway.

```typescript
// resources/assets/lib/beatmapset-search-filters.ts#135
return updateQueryString(null, { ...this.queryParams, [keyToChar[key]]: value });
```